spec/src/main/asciidoc/ch04-query-language.adoc

```diff
@@ -314,18 +314,19 @@ which the method _Character.isJavaIdentifierPart_ returns true. The
 question mark (_?_) character is reserved for use by the Jakarta
 Persistence query language.
 
-The following are reserved identifiers: ABS,
-ALL, AND, ANY, AS, ASC, AVG, BETWEEN, BIT_LENGTHfootnote:[BIT_LENGTH,
-CHAR_LENGTH, CHARACTER_LENGTH, POSITION, and UNKNOWN are not currently
-used: they are reserved for future use.],
-BOTH, BY, CASE, CHAR_LENGTH, CHARACTER_LENGTH, CLASS, COALESCE, CONCAT,
-COUNT, CURRENT_DATE, CURRENT_TIME, CURRENT_TIMESTAMP, DELETE, DESC,
-DISTINCT, ELSE, EMPTY, END, ENTRY, ESCAPE, EXISTS, EXTRACT, FALSE, FETCH,
-FROM, FUNCTION, GROUP, HAVING, IN, INDEX, INNER, IS, JOIN, KEY, LEADING,
-LEFT, LENGTH, LIKE, LOCATE, LOWER, MAX, MEMBER, MIN, MOD, NEW, NOT, NULL,
-NULLIF, OBJECT, OF, ON, OR, ORDER, OUTER, POSITION, SELECT, SET, SIZE,
-SOME, SQRT, SUBSTRING, SUM, THEN, TRAILING, TREAT, TRIM, TRUE, TYPE,
-UNKNOWN, UPDATE, UPPER, VALUE, WHEN, WHERE.
+The followingfootnote:[BIT_LENGTH, CHAR_LENGTH, CHARACTER_LENGTH,
+POSITION, and UNKNOWN are not currently used: they are reserved for
+future use.] are reserved identifiers: ABS, ALL, AND, ANY, AS, ASC,
+AVG, BETWEEN, BIT_LENGTH, BOTH, BY, CASE, CEILING, CHAR_LENGTH,
+CHARACTER_LENGTH, CLASS, COALESCE, CONCAT, COUNT, CURRENT_DATE,
+CURRENT_TIME, CURRENT_TIMESTAMP, DELETE, DESC, DISTINCT, ELSE, EMPTY,
+END, ENTRY, ESCAPE, EXISTS, EXP, EXTRACT, FALSE, FETCH, FLOOR, FROM,
+FUNCTION, GROUP, HAVING, IN, INDEX, INNER, IS, JOIN, KEY, LEADING,
+LEFT, LENGTH, LIKE, LN, LOCATE, LOWER, MAX, MEMBER, MIN, MOD, NEW,
+NOT, NULL, NULLIF, OBJECT, OF, ON, OR, ORDER, OUTER, POSITION, POWER,
+ROUND, SELECT, SET, SIGN, SIZE, SOME, SQRT, SUBSTRING, SUM, THEN,
+TRAILING, TREAT, TRIM, TRUE, TYPE, UNKNOWN, UPDATE, UPPER, VALUE,
+WHEN, WHERE.
```

Comment:
As a tangential comment, I think it's much better to _not_ include function names in the list of keywords. It's perfectly possible to parse JPQL while treating the function names as regular identifiers, just as you would do in any normal programming language.

Reply:
And I don't really know how or why `BIT_LENGTH`, `CHAR_LENGTH`, `CHARACTER_LENGTH`,
 `POSITION`, and `UNKNOWN` got onto this list (perhaps it was guff left over from EJB-QL pre 3.0), but I think they should simply be removed.

Reply:
since they are just reserved for the future use, let's keep them there for now

Reply:
I don't really care, just mentioning it because in a couple of cases I can't imagine what the future use might be: `CHAR_LENGTH`/`CHARACTER_LENGTH` in SQL are always just synonyms for JPA's `LENGTH` AFAIK.

And sure, we might in principle one day add a `POSITION` function, even though it's *really* similar to the existing `LOCATE`, but why is `POSITION` called out explicitly in that list and not `OVERLAY`?

And ... `UNKNOWN`? I can't even really speculate what that's doing there or what it might be useful for.

Anyway, it's not an important point.

Reply:
so perhaps mark them for the removal in the next update/release? I mean simply removing them right away does not look as the right thing to do.

Reply:
OK, your call.

Just removing them would not break anything though: there's no way any code can possibly be relying on them being illegal identifiers, right?

Reply:
Correct, yet it's still removal.

Reply:
Well, it's removal of a _limitation_.